code-dot-org/code-dot-org
# enable new seed process for migrated scripts

## Background

Finishes [PLAT-560]. New script serialization is already enabled for all scripts, whenever a script is created or updated. As a result, we already have .script and .script_json files for all scripts. 

## Description

This PR enables new script seeding during `rake seed:scripts` only for migrated scripts. The choice is made by parsing the .script file and looking for the is_migrated flag. The decision is made here: https://github.com/code-dot-org/code-dot-org/blob/f758fd2dd45239b20584674d51e65a6d3e485c62/dashboard/app/models/script.rb#L974-L978

## Testing story

* New unit tests ensure that the new script seeding system is invoked when is_migrated is present in the .script file. 
* Existing tests calling Script.setup help ensure that legacy scripts continue to use only the old seeding process. Because .script_json files do not exist for .script fixture files such as test-fixture.script, test-all-properties.script and test-all-fields.script, seeding for these scripts would fail if an error in our logic tried to make them use the new seeding process.
* manually verified locally that seeding restores a script to its original state after making changes to one legacy script and one migrated script.

I also manually verified that a new migrated script can be created as follows:
* /s/new --> enter script name, and press save. the legacy script editor loads
* go to dashboard-console and do `Script.find_by_name(...).update!(is_migrated: true)` (note that this step will only succeed if no levels have been added yet)
* reload the script editor page --> the new script editor loads
* save the script edit page --> updated .script and .script_json are generated, containing `is_migrated: true`

## Launch plan

This PR can't be merged until [PLAT-568] is done to verify the serialization and seed processes are working end-to-end on our entire 21-22 curriculum on an adhoc.

# Reviewer Checklist:



[PLAT-560]: https://codedotorg.atlassian.net/browse/PLAT-560

[PLAT-568]: https://codedotorg.atlassian.net/browse/PLAT-568
